core/src/main/java/org/kiwix/kiwixmobile/core/history/HistoryPresenter.kt

```diff
@@ -0,0 +1,100 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.core.history
+
+import android.util.Log
+import io.reactivex.CompletableObserver
+import io.reactivex.Observable
+import io.reactivex.Scheduler
+import io.reactivex.SingleObserver
+import io.reactivex.disposables.Disposable
+import org.kiwix.kiwixmobile.core.base.BasePresenter
+import org.kiwix.kiwixmobile.core.data.DataSource
+import org.kiwix.kiwixmobile.core.di.qualifiers.Computation
+import org.kiwix.kiwixmobile.core.di.qualifiers.MainThread
+import org.kiwix.kiwixmobile.core.history.HistoryContract.Presenter
+import org.kiwix.kiwixmobile.core.history.HistoryContract.View
+import org.kiwix.kiwixmobile.core.history.HistoryListItem.HistoryItem
+import javax.inject.Inject
+
+internal class HistoryPresenter @Inject constructor(
+  private val dataSource: DataSource,
+  @param:MainThread private val mainThread: Scheduler,
+  @param:Computation private val computation: Scheduler
+) : BasePresenter<View>(), Presenter {
+  private var disposable: Disposable? = null
+
+  override fun loadHistory(showHistoryCurrentBook: Boolean) {
+    dataSource.getDateCategorizedHistory(showHistoryCurrentBook)
+      .subscribe(object : SingleObserver<List<HistoryListItem>> {
```

Comment:
Use the lambda version of subscribe. Don't use object.

Reply:
Tried doing that but couldn't get it to work. It says 'Overload resolution ambiguity' when doing this: 
```
.subscribe({
        override fun onSubscribe(d: Disposable) {
          if (disposable != null && !disposable!!.isDisposed) {
            disposable!!.dispose()
          }
          ...
})
```
Could you point me closer to how I should implement an interface using lambda?

Reply:
you don't declare a function name override.

`subscribe(
{/*this is the on next*/},
{/*this is the on error*/},
{/*this is the on complete*/})`

Reply:
ah, cool. Thanks!

Reply:
subscribe has 1,2 & 3 param versions so if you don't care about onComplete/Error you can just use. That is also advice for a Flowable which represents multiple values over time, this is a Single which represents 1 value so it doesn't have a version with onComplete as it either goes to `onSuccess/Error`

![image](https://user-images.githubusercontent.com/3358282/78796574-6ab8fc80-79ae-11ea-81b1-6a7f50361d39.png)

The type signatures help a lot here but there are a lot of methods

Reply:
This is what I did and seems to work ok but it is missing the disposable handling part. So I can't use the onSubscribe without using a SingleObserver?

```
dataSource.getDateCategorizedHistory(showHistoryCurrentBook)
      .subscribe(
        {histories: List<HistoryListItem> -> view?.updateHistoryList(histories)},
        {e:Throwable -> Log.e("HistoryPresenter", "Failed to load history.", e)})
 

```

Reply:
`.subscribe` returns the disposable, you can add that disposable to the compositeDisposable.
I guess we can avoid `switchMap` stuff by simply doing
```
disposable?.takeIf{ !it.isDisposed }.let { it.dispose }
disposable = blahblahblah.subscribe()
compositeDisposable.add(disposable)
```

Reply:
That's lots better! thanks.